README.md
# typed-encoding
Type level annotations, string transformations, and other goodies that make programming strings safer.

## Motivation
I have recently spent a lot of time troubleshooting various `Base64`, `quoted-printable`, and `UTF-8` encoding issues.  
I decided to write a library that will help avoiding issues like these.

This library allows to specify and work with types like

```Haskell
-- some data encoded in base 64
mydata :: Enc '["enc-B64"] c ByteString

-- some text (utf8) data encoded in base 64 
myData :: Enc '["enc-B64", "r-UTF8"] c ByteString
```

It allows to define precise string content annotations like:

```Haskell
ipaddr :: Enc '["r-IpV4"] c Text
```

and provides ways for 

- encoding
- decoding
- recreation (encoding validation)
- type conversions
- converting types to encoded strings
- typesafe conversion of encoded strings to types

Partial and dangerous things like `decodeUtf8` are no longer dangerous, 
`ByteString` `Text` conversions become fully reversible.  Life is good!

... but this approach seems to be a bit more...

```Haskell
-- upper cased text encoded as base64
example :: Enc '["enc-B64", "do-UPPER"] () T.Text
example = encodeAll . toEncoding () $ "some text goes here"
```

It becomes a type directed, declarative approach to string transformations.

Transformations can be

- used with parameters
- applied or undone partially (if encoding is reversible)

One of more interesting uses of this library are encoding restrictions included in this library.   
Example are (arbitrary) bounded alpha-numeric (`"r-ban"`) restrictions.

```Haskell
-- allow only properly formatted phone numbers

type PhoneSymbol = "r-ban:999-999-9999"
phone :: Enc '[PhoneSymbol] () T.Text
phone = ... 
```

The author often uses _typed-encoding_ with _Servant_ (`HttpApiData` instances are not included), e.g.:

```Haskell
type LookupByPhone = 
  "customer"
  :> "byphone"
  :> Capture "phone" (Enc '[PhoneSymbol] () T.Text)
  :> Get '[JSON] ([Customer])
```

or to get type safety over text document using Unix vs Windows line breaks!


## Goals and limitations

The main goal is to provide improved type safety for programs that use string encodings and 
transformations.  _Not to provide encoding implementation type safety_. 
_Encoding and string manipulation libraries are typically well established and tested, type safety is really needed at the usage site, not at the implementation site_.

This library approach is to fight issues with (value level) strings using (type level) strings. Using `Symbol`-s effectively forces us to play the orphan instances game.   
One of the long term goals is for this library to provide combinator alternatives to typeclass polymorphism so that the orphan instances are more of a convenience and not the necessity.  


## Examples 

Here are some code examples:

- [Overview](src/Examples/TypedEncoding/Overview.hs)
- [Conversions between encodings](src/Examples/TypedEncoding/Conversions.hs)
- [DIY encoding, error handling](src/Examples/TypedEncoding/Instances/DiySignEncoding.hs)
- [To and from string conversions](src/Examples/TypedEncoding/ToEncString.hs)
- [Unsafe - working inside encodings](src/Examples/TypedEncoding/Unsafe.hs)
 

## Hackage

https://hackage.haskell.org/package/typed-encoding

## Other encoding packages

My approach will be to write specific encodings (e.g. _HTTP_) or wrap encodings from other packages using separate "bridge" projects.

Currently /typed-encoding/ depends on

- /base64-bytestring/ because it was my driving example, this is likely to move out to a separate bridge project at some point. 

Bridge work:

- [typed-encoding-encoding](https://github.com/rpeszek/typed-encoding-encoding) bridges [encoding](https://github.com/dmwit/encoding) package 


## Tested with

- stack (1.9.3) lts-14.27 (ghc-8.6.5)
- stack (2.5.1) lts-16.27 (ghc-8.8.4)
- needs ghc >= 8.2.2, base >=4.10 for GHC.TypeLits support

## Known issues

- running test suite: cabal has problems with doctest, use stack  
   https://github.com/haskell/cabal/issues/6087   


ChangeLog.md
# Changelog for typed-encoding

## Anticipated future breaking changes

- `Data.TypedEncoding.Common.Class.IsStringR` expected to be be changed / replaced
- (post 0.5) "enc-B64" could be moved to a different package (more distant goal)


## 0.5.2.3
- compile with ghc 9.4.7 / lts-21.14
## 0.5.2.2

- `instance UnexpectedDecodeErr Identity` does not use `fail` to allow for base >= 4.9 compilation
## 0.5.2.1

- documentation improvements
## 0.5.2.0

- Fixed up IsSuperset type family improving transitivity
- added `_mkEncoding1` `runEncoding1'` functions that show equivalence of `Encoding` to one level encoding
- added convenience one level encoding conversion functions
- base64-bytestring deps relaxed

## 0.5.1.0

- "r-B64" added
- @implVerifyR@ convenience function added to 'Data.TypedEncoding.Instances.Support.Encode'

## 0.5.0

### Changes on a high level

- Most of the changes should not create a big impact on upgrading.  Many definitions were moved to a different module but these modules are and had been re-exported by either `Data.TypedEncoding` or
`Data.TypedEncoding.Instances.Support`
- Some functionality has been moved to Examples or removed, notably:
  - "do-" encodings
  - `SomeEnc`, `SomeAnnotation`
  - `HasA` typeclass
- Some functions have been renamed or type signatures adjusted to follow consistent naming conventions.  In most cases the changes have been made on previously deprecated definitions.


###  Details 

- Data.TypedEncoding.Instances.Do.Sample moved to Examples
- Examples.TypedEncoding folder re-org
- `Data.TypedEncoding.Instances.Support.Helpers` removed `foldEncStr`, `foldCheckedEncStr`
   renamed  `splitSomePayload` to `splitCheckedPayload`
- `HasA` typeclass moved to Examples
- removed experimental `Data.TypedEncoding.Instances.Restriction.Bool` in favor of combinator helpers
  `Data.TypedEncoding.Instances.Support.Bool`
- `Data.TypedEncoding.Common.Types.SomeEnc` moved to Examples  
- `Data.TypedEncoding.Common.Types.SomeAnnotation` moved to Examples
- camel-case of some property names
- Text instances for "Base64" moved to `Data.TypedEncoding.Instances.Enc.Warn.Base64`
- Removed instanced for `"r-()"` encoding
- Functions from `Data.TypedEncoding.Instances.Support.Common` moved to `Data.TypedEncoding.Instances.Support.Decode`
- Signature changed in previously deprecated function `runDecoding` to match `mn ~ alg` convention and deprecation removed
- Signature changed in previously deprecated function `runDecodings` to match `mns ~ algs` convention and deprecation removed
-  Signature changed in previously deprecated function `runValidation` to match `mns ~ algs` convention and deprecation removed
- `runValidationChecks` renamed to `runValidationChecks'` to match /typed-encoding/ naming conventions. 
- removed deprecated `propEncodesInto'`
- moved `Append` type family from from `Data.TypedEncoding.Common.Class.Util` to `Data.TypedEncoding.Common.Util.TypeLits`
- `Data.TypedEncoding.Common.Class.Util` renamed to `Data.TypedEncoding.Common.Class.Common`
- function `extractEither` removed from `Data.TypedEncoding.Internal.Util`
- function `withSomeSymbol` moved to `Data.TypedEncoding.Common.Util.TypeLits`
- function `proxyCons` moved to `Data.TypedEncoding.Common.Util.TypeLits`

- More general instances for some encodings in `Data.TypedEncoding.Instances.Restriction.Misc`
- `mkDecoding` deprecated in favor of `_mkDecoding` to follow the naming convention
- `mkValidation` deprecated in favor of `_mkValidation` to follow the naming convention
- `validR'` function renamed to `_validR`

## 0.4.2

- `Data.TypedEncoding.Instances.Support.Bool` combinators ("r-bool" remains experimental)
- Correction in `r-ban` error message typo ("'G' not bounded by 'F'")
- `_implEncFromString` creates "r-" encoding from `FromEncString` definition
- property changes: `propEncodesIntoCheck` added; `propSuperset'`,  `propEncodesInto'` deprecated

## 0.4.1

- Code Changes. Backward compatible
   - changed order in `IsSuperset` definition to speed up compilation of of more common cases (there is a small chance that it impacts GHC error messages)
   - Faster "r-UTF8", possible issue is changed error message in case ByteString is invalid.
   - Deprecated 'validFromEnc' for its confusing name
   - Deprecated `runDecodings` in favor of consistently named `runDecodings'`
   - Deprecated `runDecoding` in favor of consistently named `runDecoding'`
   - Deprecated `runValidation` in favor of consistently named `runValidation'`

- Documentation / code comment fixes and improvements.

- New functionality `validRFromEnc'` replacing 'validFromEnc' confusing name
  - `propCompEncoding` property
  - `propSupersetCheck` property
  - `propSafeDecoding'` properties
  - `propSafeValidatedDecoding` properties
  - `IsEnc` type family and `Encoding` constraint
  - `getUncheckedPayload` function
  - `pack` and `unpack` overloads in `Data.TypedEncoding.Conv.ByteString.Char8`

- Fixes
  - corrected `propEncodesInto'` property test specification


## 0.4

- Breaking 
  - IsSupersetOpen type family type arguments have changed

- Potentially Breaking  
  (These changes should be backward compatible in almost all cases):
  - Stronger (more precise) constraints on all functions `Data.TypedEncoding.Conv`
  - Compilation errors emitted from `IsSuperset` are different
  - "r-ban" now only allows ASCII chars in annotation name, errors-out otherwise
   
- New
  - `"r-CHAR8"` phantom restriction and `Superset` modified for "r-CHAR8"
  - `"r-UNICODE.D76"` /text/ character set restriction and `Superset` modifications
  - `Superset` constraint added back (different than in 0.2)
  - properties for `Superset` testing
  - `"r-ByteRep"` annotation used as a marker of low level use of `Char` instead of `Word8` for `ByteString` work.

- Improved:
  - `Data.TypedEncoding.Conv` `Text`, `String` and `ByteString` conversions are now more type safe and less error prone.

src/Data/TypedEncoding.hs

{-# LANGUAGE TypeApplications #-}
{-# LANGUAGE ScopedTypeVariables #-}
{-# LANGUAGE TypeFamilies #-}
{-# LANGUAGE DataKinds #-}
-- |
-- = Overview
--
-- This library uses 'GHC.TypeLits' symbols to specify and work with types like
--
-- @
-- -- Base 64 encoded bytes (could represent binary files)
-- Enc '["enc-B64"] () ByteString
--
-- -- Base 64 encoded UTF8 bytes
-- Enc '["enc-B64", "r-UTF8"] () ByteString
--
-- -- Text that contains only ASCII characters
-- Enc '["r-ASCII"] () Text
-- @
--
-- or to do transformations to strings like
--
-- @
-- upper :: Text -> Enc '["do-UPPER"] c Text
-- upper = ...
-- @
--
-- or to define precise types to use with 'toEncString' and 'fromEncString'
-- 
-- @
-- date :: Enc '["r-date-%d/%b/%Y:%X %Z"] () Text
-- date = toEncString ...
-- @
--
-- Primary focus of /type-encodings/ is to provide type safe
--
-- * /encoding/
-- * /decoding/
-- * /validation (recreation)/ (verification of existing payload)
-- * safe type conversions between encoded types
-- * combinators for creating new encodings from existing encodings (e.g. by applying Boolean rules)
--
-- of string-like data (@ByteString@, @Text@) that is subject of some
-- encoding or formatting restrictions.
--
-- as well as
--
-- * /toEncString/ 
-- * /fromEncString/ 
--
-- conversions.
--
-- = Groups of annotations
--
-- typed-encoding uses type annotations grouped into semantic categories
--
-- == "r-" restriction / predicate
--
-- * /encoding/ is a partial identity
-- * /validation/ is a partial identity (matching encoding)
-- * /decoding/ is identity
--
-- Examples: @"r-UTF8"@, @"r-ASCII"@, upper alpha-numeric bound /r-ban/ restrictions like @"r-ban:999-999-9999"@
--
-- == "do-" transformations 
--
--(not provided in this library other than as /Examples/ "Examples.TypedEncoding.Instances.Do.Sample")
--
-- * /encoding/ applies transformation to the string (could be partial)
-- * /decoding/ - typically none
-- * /validation/ - typically none but, if present, verifies the payload has expected data (e.g. only uppercase chars for "do-UPPER")
--
-- Examples: @"do-UPPER"@, @"do-lower"@, @"do-reverse"@
--
-- == "enc-" data encoding that is not "r-"
--
-- * /encoding/ applies encoding transformation to the string (could be partial)
-- * /decoding/ reverses the transformation (can be now be used as pure function)
-- * /validation/ verifies that the payload has correctly encoded data
--
-- Examples: @"enc-B64"@
-- 
--
-- = Call Site Usage
--
-- To use this library import this module and one or more /instance/ or /combinator/ module.
--
-- Here is list of instance modules available in typed-encoding library itself
--
-- * "Data.TypedEncoding.Instances.Enc.Base64"
-- * "Data.TypedEncoding.Instances.Restriction.Misc" (replaces @Common@ from v0.2)
-- * "Data.TypedEncoding.Instances.Restriction.ASCII" 
-- * "Data.TypedEncoding.Instances.Restriction.UTF8" 
-- * "Data.TypedEncoding.Instances.Restriction.BoundedAlphaNums" (moved from @Combinators@ to @Instances@ in v0.3)
-- 
-- ... and needed conversions. 
--
-- Conversion combinator module structure is similar to the one found in /text/ and /bytestring/ packages.
--
-- Conversion is /typed-encoding/ are safe and reversible!
--
-- Please see comments in 
--    
-- * "Data.TypedEncoding.Conv" 
--
-- for more information.
--
-- The instance list is not intended to be exhaustive, rather separate libraries
-- can provide instances for other encodings and transformations.
--
-- = New encoding instance creation
--
-- To implement a new encoding import
--
-- * "Data.TypedEncoding.Instances.Support"
--
-- = Examples
--
-- Examples of how to use this library are included in
--
-- * "Examples.TypedEncoding"    

module Data.TypedEncoding (
  
    -- * @Enc@ and basic combinators
    Enc
    , toEncoding
    , fromEncoding
    , getPayload
  
    -- * Untyped versions of @Enc@
    , module  Data.TypedEncoding.Common.Types.CheckedEnc

    -- * @Encoding@ and basic combinators
    , Encoding (..)
    , _mkEncoding
    , _mkEncoding1 
    , runEncoding'
    , _runEncoding 
    , runEncoding1'
  
    -- * List of encodings
    , Encodings (..)
    , runEncodings'
    , _runEncodings

    -- * Similar to @Encoding@ and @Encodings@ but cover /Decoding/ and /Validation/
    , module Data.TypedEncoding.Common.Types.Decoding
    , module Data.TypedEncoding.Common.Types.Validation

    -- * @UncheckedEnc@ is an /untyped/ version of Enc that represents not validated encoding      
    , module Data.TypedEncoding.Common.Types.UncheckedEnc

    -- * Laws / properties

    , propSafeDecoding' 
    , _propSafeDecoding
    , propSafeValidatedDecoding'
    , _propSafeValidatedDecoding

    -- * Encoding Classes
    , Encode (..)
    , EncodeAll (..)

    -- * Decoding, Validation and Other Classes
    , module Data.TypedEncoding.Common.Class
  
      -- * Combinators
    , module Data.TypedEncoding.Combinators.Common
    , module Data.TypedEncoding.Combinators.Encode
    , module Data.TypedEncoding.Combinators.Decode
    , module Data.TypedEncoding.Combinators.Validate
    , module Data.TypedEncoding.Combinators.Unsafe
    , module Data.TypedEncoding.Combinators.ToEncStr
    , module Data.TypedEncoding.Combinators.Promotion

    -- * Exceptions 
    , module Data.TypedEncoding.Common.Types.Exceptions

    -- * Other
    , module Data.TypedEncoding.Common.Types.Common

 ) where

import           Data.TypedEncoding.Common.Types.Enc
import           Data.TypedEncoding.Common.Types.Decoding
import           Data.TypedEncoding.Common.Types.Validation

import           Data.TypedEncoding.Common.Types.Common
import           Data.TypedEncoding.Common.Types.CheckedEnc
import           Data.TypedEncoding.Common.Types.UncheckedEnc
import           Data.TypedEncoding.Common.Types.Exceptions
import           Data.TypedEncoding.Common.Class
import           Data.TypedEncoding.Combinators.Common
import           Data.TypedEncoding.Combinators.Encode
import           Data.TypedEncoding.Combinators.Decode
import           Data.TypedEncoding.Combinators.Validate
import           Data.TypedEncoding.Combinators.Unsafe
import           Data.TypedEncoding.Combinators.ToEncStr
import           Data.TypedEncoding.Combinators.Promotion


-- | 
-- Main property that encodings are expected to enforce.
--
-- Decoding is safe and can use @Identity@ instance of 'UnexpectedDecodeErr'.
--
-- Errors are handled during the encoding phase.
propSafeDecoding' :: forall alg nm c str. (Eq c, Eq str) =>
                     Encoding (Either EncodeEx) nm alg c str 
                     -> Decoding (Either UnexpectedDecodeEx) nm alg c str 
                     -> c 
                     -> str 
                     -> Bool
propSafeDecoding' encf decf c str = Right enc0 == edec
   where 
       enc0 = toEncoding c str
       eenc = runEncoding' @alg encf enc0
       edec = case eenc of 
           Right enc -> either (Left . show) Right $ runDecoding' @alg decf enc
           Left enc -> Right enc0 -- quick and dirty, magically decode all encoded failures  


_propSafeDecoding :: forall nm c str alg . (Algorithm nm alg, Eq c, Eq str) => 
                     Encoding (Either EncodeEx) nm alg c str 
                     -> Decoding (Either UnexpectedDecodeEx) nm alg c str 
                     -> c 
                     -> str 
                     -> Bool
_propSafeDecoding = propSafeDecoding' @(AlgNm nm)

-- |
-- Similar to 'propSafeDecoding'' but 'Validation' based.
-- 'Validation' acts as 'Decoding' recovering original payload value.
-- Recovering with validation keeps the encoded value and that value
-- is supposed to decode without error. 
--
-- Expects input of encoded values
propSafeValidatedDecoding' :: forall alg nm c str. (Eq c, Eq str) =>
                     Validation (Either RecreateEx) nm alg c str 
                     -> Decoding (Either UnexpectedDecodeEx) nm alg c str 
                     -> c 
                     -> str 
                     -> Bool
propSafeValidatedDecoding' validf decf c str = edec == echeck
   where 
       enc0 = toEncoding c str
       valfs = validf `ConsV` ZeroV
       eenc = recreateWithValidations @'[alg] @'[nm] valfs enc0 
       (edec :: Either String str, echeck :: Either String str) = case eenc of 
           Right enc -> (either (Left . show) (Right . getPayload) $ runDecoding' @alg decf enc
                         , either (Left . show) (Right . getPayload) $ runValidation' @alg validf enc)
           Left _ -> (Left "not-recovered", Left "not-recovered")  

_propSafeValidatedDecoding :: forall nm c str alg. (Algorithm nm alg, Eq c, Eq str) =>
                     Validation (Either RecreateEx) nm alg c str 
                     -> Decoding (Either UnexpectedDecodeEx) nm alg c str 
                     -> c 
                     -> str 
                     -> Bool
_propSafeValidatedDecoding = propSafeValidatedDecoding' @(AlgNm nm)